### PR TITLE
Installer | Remove "dialog" from Debian/Ubuntu installer deps

### PR DESCRIPTION
**Pull request by MichaIng**
- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Signed-off-by: MichaIng <micha@dietpi.com>

---
**What does this PR aim to accomplish?:**
+ The installer uses `whiptail`, thus `dialog` is not required.

**How does this PR accomplish the above?:**
+ `whiptail` is removed from Debian/Ubuntu installer deps variable.
+ It does that only for Debian/Ubuntu, since I have no chance to test on rpm-based distros. Also on those, the `whiptail` package is not in installer deps, so I am wondering if this is missing, or part of those OS's core packages, or included in the `dialog` package?

**What documentation changes (if any) are needed to support this PR?:**
+ None

### CODE REVIEW

**Comment by MichaIng** [comment]
If someone has a chance to test on supported rpm-based distros, I can add this as well. And probably more important, isn't the `whiptail` package missing there?

And one minor issue I can think about:
- One has Pi-hole installed prior to this PR/release, so dialog is installed as well.
- One updates Pi-hole to contain this PR.
- One uninstalls Pi-hole and chooses to uninstall deps as well.
- `dialog` is now missing from deps, so it is not asked whether to uninstall it.
- However with the current method (sourcing installer script), there is no chance to go through the dependency list of time/version of initial Pi-hole install, and it is not a big problem as well.

**Comment by technicalpyro** [comment]
We have found through our Testing that `dialog` or rather lack thereof on many supported OS iso's generally the ones VPS providers use that the installer would fail without it that was why the `dialog` dependency was added.  

How does your PR address this issue?

**Comment by MichaIng** [comment]
@technicalpyro 
Can you link issues or test results where this was an issue? Actually that should not be possible:
- [whiptail](https://packages.debian.org/buster/whiptail) and [dialog](https://packages.debian.org/buster/dialog) are two entirely independent console-based input dialog solutions. On Ubuntu these packages + dependencies are exactly the same btw.
- As long as only the `whiptail` binary/command is called, which is the case with Pi-hole installer, `dialog` is just obsolete/not used, how should it.
- Also both packages don't have any recommends or suggestions, just dependencies, so there is no somewhat shared library that might be pulled by the one package, and not pulled automatically, but used by the other one if present. That was what I had in mind first, when reading you comment.
- So even if some VPS or other kind of system does for some reason not support `whiptail`, since the installer script has no `dialog` command fallback implemented, it would just fail, while `dialog` will still be unused.
- And about the support, on DietPi all interactive scripts make vast use of whiptail, and only whiptail, and it is used on nearly all ARMs on the marked, x86_64 PCs, VMs and VPS, and I am not aware of a single report that `whiptail` failed. Of course a TTY must be attached, so the script must be called from an interactive shell, but that is true for `dialog` as well. And the script requires interactive input anyway, besides it is called without `--unattended` which skips all `whiptail` calls then.

So far about the theory, but probably I've overseen something. I would then consider it as whiptail-related bug which would be interesting to investigate. Even more interesting then, how the presence of the dialog binary, without being called, can fix it 🙂.

**Comment by technicalpyro** [comment]
original issue report is #1278 and fix was #1280

**Comment by MichaIng** [comment]
@technicalpyro 
Okay, it is not the installers whiptail dialogs, but `debconf-apt-progress`. Actually it does not require any of both, it tries `whiptail` first, falls back to `dialog` otherwise, falls back to perl `Readline` otherwise and uses `Teletype` if non of the above is available:
```
# debconf-apt-progress -- apt-get -y install net-tools
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC contains: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.24.1 /usr/local/share/perl/5.24.1 /usr/lib/x86_64-linux-gnu/perl5/5.24 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.24 /usr/share/perl/5.24 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base .) at /usr/share/perl5/Debconf/FrontEnd/Readline.pm line 7.)
debconf: falling back to frontend: Teletype
Installing packages
-------------------

..15%
..29%
..43%
..57%
..100%
```
- As fast as you install whiptail, it is used by debconf.

This also fits the `debconf` package dependencies: https://packages.debian.org/buster/debconf
- It does not depend on any of those frontend packages, those are suggestions only to make it more beautiful.

Ah you know what:
> Checking for whiptail... added to install list!

Of course if you use debconf already for those installer dependencies, both, whiptail and dialog might not yet been present. So actually #1280 would not fix anything, since the issue occurs before the fix (dialog install) has been done 😄.

In your case, debconf went on to use the Readline fallback frontend and starts to install packages. Not sure why it stopped there, or did you simply not wait long enough? Probably this was an issue with either Readline or how debconf invokes it:
```
# debconf-apt-progress -- apt-get -y install net-tools
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76.)
debconf: falling back to frontend: Readline
Installing packages
-------------------

..15%
..29%
..43%
..57%
..100%
```
- Strange since this looks exactly like the Teletype frontend above.
- 🈴 The problem indeed is that it does not proceed afterwards. When calling this from console, I need to interrupt it, I guess it would hang the script for the same reason. However, it doesn't matter, see below...
________________
The major issue, that is indeed a bug/quirk in installer code, is that whiptail (and/or dialog) is installed with a method that depends on whiptail or dialog, otherwise at best would fallback to ugly frontends with error messages with one there that would even hang: https://github.com/pi-hole/pi-hole/blob/master/automated%20install/basic-install.sh#L1593-L1634
- #1280 does not solve this issue, the same like having `whiptail` in the installer deps did not prevent it in the first place.
- `whiptail` must simply be installed outside of `debconf-apt-progress`, then one does not require dialog for any further install.

**Comment by MichaIng** [comment]
@dschaper 
Just to pull in someone and sum up:
- `whiptail` and/or `dialog` cannot be installed via `debconf-apt-progress`, since `debconf-apt-progress` uses either of those two to visualize the package install.
- If `whiptail` and `dialog` are both not installed, but `perl` is, debconf will fallback to the `Readline` frontend, which indeed seems to hang scripts (an external issue that probably should be reported to debconf and/or perl devs). Since `whiptail` and `dialog` are installed currently via `debconf-apt-progress`, the hang happens before the assumed fix #1280 has any effect.
- Since `whiptail` IS used, if present, `dialog` is still never required, but `whiptail` instead must be installed outside of `debconf-apt-progress` wrapper, to have debconf using it afterwards.
- The same is actually true for the `debconf` package itself of course :smile:.
- I can add it to this PR, but not sure how you like it: `apt-get -qq install debconf whiptail`? This has least output and `-qq` includes `-y`. STDOUT/STDERR to /dev/null is not what I would do. STDOUT allows to see progress, otherwise on slower systems/network there would be no hint for progress. STDOUT should always be shown to allow debugging in case of error. What you think?

I am not sure about the rpm-based installer, which frontends they use, but AFAIK whiptail there is even more standard then on Debian/Ubuntu, so dialog should be even less used by any install progress visualizer. Does it have an own package, or is it essentail, thus can be assumed to always be present?

**Comment by bcambl** [comment]
I don't think a dependency should be removed unless the original issue where the dependency was added has been tested.   
FWIW `dialog` is only ~234k size on disk and should not cause any issues to be installed as a failsafe.

**Comment by MichaIng** [comment]
@bcambl 
Of course it must be tested. However it seems that the fix itself was not tested probably that time 😉. If I am not completely wrong with some logic, then as said, nothing has been fixed with this. I guess it was tested that way:
- Installer failed
- Install dialog
- Installer succeeded
- "Fix" implemented

But installing `dialog` from within the script AFTER it is called by `debconf` of course cannot work.
And the assumption, that `dialog` itself is the fix, is as well the wrong assumption, since installing `whiptail` before running the installer would have worked the same way.

So long story short. Removing a dependency or not, the initial issue must be fixed anyway, since it is not. `whiptail` and `debconf` must be installed outside and before using `debconf-apt-progress`.

I will implement this tomorrow, currently too busy with other stuff.

**Comment by MichaIng** [comment]
Okay guys, on RPM systems neither `dnf` nor `yum` (and how they are implemented in the installer) make no use of whiptail or dialog, so the `dialog` dependency makes even less sense there. For `whiptail`, which is used directly by the installer, it's more complicated:
- Arch Linux: A `whiptail` package does not exist, the [libnewt](https://www.archlinux.org/packages/community/x86_64/libnewt/) package contains the `whiptail` binary.
- Fedora: As well no whiptail pacakge available, but [newt](https://koji.fedoraproject.org/koji/rpminfo?rpmID=18733826), which provides the `whiptail` binary.
- On CentOS it's the same: `newt`
- Best overview I could find: https://pkgs.org/download/newt
- So I guess, it makes sense to search for the `newt` package first, and if it's not available for `libnewt`. However this is not something I gonna implement as I have no RPM system to test with.

So I will add the following now to really fix the case of failing/hanging installer on Debian, if no dialog or whiptail is installed before, and another issue, where installer still fails if no `debconf` was installed before, although this should be very rare.

**Comment by bcambl** [comment]
Sorry for the slow response on this.. I have been busy with my job and personal life recently. Thank you for the PR as it does have some valid changes, but please know this is not considered high priority due to being more of a _clean-up_ than a _bug fix_. Please do not be offended if this PR is slow. That said, I did find a semi-related bug while reviewing this which will be created as a separate issue.

Re: CentOS/Fedora:
- Yes, `newt` is the packages which supplies the required libs on the two rpm based distros we support.
- Yes, `dialog` is not required and can be removed for the two rpm based distros we support.

My thoughts were to re-order the dependency arrays to put `whiptail` and `newt` at index 0 of the arrays. I noticed that the arrays were originally created in alphabetical while additional dependencies were simply appended. 

Arch is not one of our supported distros with the current state of the installer and we are not planning to add more distros to the current installer script due to future plans.

So you may update this PR by moving `dialog` from the Fedora/CentOS dep array, and move `whiptail` and `newt` to the front of the arrays.

**Comment by bcambl** [comment]
Please revert your second commit and only make the changes I have requested to `INSTALLER_DEPS`

**Comment by MichaIng** [comment]
@bcambl 
Jep I just committed the removal of dialog for RPM. As well I fixed the initial issue (why dialog was added) by creating a separate APT_DEPS array that hold the three packages that are required to assure successful `debconf-apt-progress` calls. It is not too beautiful but as long as you want to keep the progress bar, it is the only way.

I am btw not sure if `apt-utils` is required. AFAIK this is only used by debconf to pre-configure packages, which is nothing that Pi-hole actively does or require? However This is really a separate topic and should be not part of this PR as well 😉.

> put whiptail and newt at index 0

Just having it the first array entry does not fix things, but ready the commit/code comments at best 😉. In case of `newt` it does not play a role, since the RPM package tools to not make use of it.

**Comment by bcambl** [comment]
`newt` provides `whiptail` on the rpm distros.  
I requested you to revert your second commit because it contained changes outside the scope of this PR (indentation). I do not want to implement your `APT_DEPS` solution as I would like to discuss with the team more about the usage of `debconf-apt-progress`.

I must sign off now, cheers.

**Comment by MichaIng** [comment]
@bcambl 
Finally I think both belongs into a single PR since it is about correcting a fix for a still existing issue. Removing dialog reverts the old false fix and the `APT_DEPS` part is the correct fix. What you think?

But I agree/understand that the implementation is something that you want to discuss with the team. As you said, dialog is a small package, so there is no time pressure to remove this now/quickly without implementing a real fix for the initial issue.

**Comment by bcambl** [comment]
#2962 created for the removal of `debconf-apt-progress`.  
Closing.
